### PR TITLE
Simplifier: new interface for arithmetic expressions

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 48e0755673f9934fcdbb5c3313430c887ea5fb7f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117316535

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c43a6348fa64e06c1ce65f77b45396b356e5583b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117585354

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -2013,11 +2013,10 @@ simplify_exprt::simplify_byte_extract(const byte_extract_exprt &expr)
   {
     auto tmp = expr;
 
-    tmp.offset() =
-      plus_exprt(to_byte_extract_expr(expr.op()).offset(), expr.offset());
-    simplify_plus(tmp.offset());
-
+    tmp.offset() = simplify_plus(
+      plus_exprt(to_byte_extract_expr(expr.op()).offset(), expr.offset()));
```
Comment (allredj) [inline src/util/simplify_expr.cpp:2004]: :pick: Should be in separate commit

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -2013,11 +2013,10 @@ simplify_exprt::simplify_byte_extract(const byte_extract_exprt &expr)
...
     tmp.op() = to_byte_extract_expr(expr.op()).op();
+
```
Comment (allredj) [inline src/util/simplify_expr.cpp:2006]: :pick: Should be in separate commit

File: src/util/simplify_expr.cpp [hunk 3]
```diff
@@ -2575,25 +2574,81 @@ bool simplify_exprt::simplify_node(exprt &expr)
...
   else if(expr.id()==ID_mult)
-    no_change = simplify_mult(expr) && no_change;
+  {
+    auto r = simplify_mult(to_mult_expr(expr));
+    if(r.has_changed())
+    {
+      no_change = false;
+      expr = r.expr;
+    }
+  }
```
Comment (allredj) [inline src/util/simplify_expr.cpp:2671]: That's a lot of code duplication. Should be factored out.

**Review by allredj — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1c5dd94bf7f5dbd6e90172e9c6ad98d8fe81663f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117643436

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:2004]: That won't work, as this change requires the change to the signature of `simplify_plus` (i.e., the commit wouldn't compile by itself).

File: src/util/simplify_expr.cpp (hunk 3 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:2671]: Yes, will happen once majority of calls uses the new interface.
